ships/0028-buildrun-cleanup.md

```diff
@@ -0,0 +1,175 @@
+
+We will introduce a new `build-limit-cleanup-controller` which will reconcile Builds. It will watch Builds with those predicate functions:
+
+- Create: if the Build has limits configured. This is handled for managers becoming the leader.
+- Update: if the Build has limits introduced, or changed to lower values.
+
+It will watch BuildRuns with those predicate functions:
+
+- Update: if the BuildRun completed (`status` of condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), and if it is related to a Build (standalone BuildRuns when introduced with SHIP 0029 will be ignored).
+
+In the `Reconcile` function, we will check the Build: if it has any limit set, then
+
+- We will run a list for BuildRuns in the namespace with the label filter `build.shipwright.io/name: <BUILD_NAME>`.
+- All BuildRuns will be put into three buckets: succeeded (`status` of condition `type=Succeeded` set to `"true"`), failed (`status` of condition `type=Succeeded` set to `"false"`) and the rest.
+- If the number of succeeded BuildRuns is larger than `succeededLimit`, then the oldest succeeded BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+- If the number of failed BuildRuns is larger than `failedLimit`, then the oldest failed BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+
+### Staging
+
+The implementation can be done in stages. In particular, the TTL-based cleanup and the limit-based cleanup are very much independent.
+
+### Documentation updates
+
+We will need to update the documentation about builds to introduce this new feature.
+
+### User Stories
+
+#### As a Build user, I want to define how many finished BuildRuns I want to keep for my Build so that the amount of finished BuildRuns does not grow endlessly
+
+#### As a Build user, I want to define how long a finished BuildRun should stay so that I do not have obsolete and very old BuildRuns in the system
+
+### Test Plan
+
+It should include:
+
+- Integration tests
+- Unit tests
+- e2e tests
+
+### Release Criteria
+
+Should be available before release v1.0.0
+
+### Risks and Mitigations
+
+We will have more controllers and reconcilers. Consumers may have to increase memory limits in the deployment.
+
+For the support of limits, especially the required `list` calls can become a problem if `succeededLimit`, or `failedLimit` are set to high values and we have to retrieve a large amount of BuildRuns from the API server and have them in memory. A mitigation would be that we define maximum values for those values (maybe 100).
```

Comment:
Perhaps we enforce an upper bound on succeeded + failed limits in the main `Build` controller and/or an additional webhook? Something to consider as a future enhancement.